lib/sparse/src/index/search_context.rs

```diff
@@ -247,30 +244,49 @@ impl<'a> SearchContext<'a> {
             return Vec::new();
         }
         let mut best_min_score = f32::MIN;
-        while let Some(candidate) = self.advance() {
-            // check filter condition
-            if !filter_condition(candidate.idx) {
-                continue;
-            }
-            // check for cancellation
+        loop {
+            // check for cancellation (atomic amortized by batch)
             if self.is_stopped.load(Relaxed) {
                 break;
             }
-            // push candidate to result queue
-            self.result_queue.push(candidate);
+
+            // prepare next iterator of batched ids
+            let start_batch_id = match self.min_record_id {
+                Some(min_id) => min_id,
+                None => break, // all posting lists exhausted
+            };
+
+            // compute batch range of contiguous ids for the next batch
+            let last_batch_id = min(
+                start_batch_id + ADVANCE_BATCH_SIZE as u32,
+                self.max_record_id,
+            );
+            let batch_len = last_batch_id - start_batch_id + 1;
+
+            // init batch scores
```

Comment:
I propose to move this part
```
            // init batch scores
            self.batch_scores.clear();
            self.batch_scores.resize(batch_len as usize, 0.0);
```
into the `advance_batch` method. Because `self.batch` is `advance_batch` internal cache and we don't use `self.batch` inside search function directly